app/controllers/assignments_controller.rb

```diff
@@ -577,74 +582,76 @@ def download_config_files
-      flash_message(:error, I18n.t('upload_errors.blank'))
-    end
-    if File.extname(upload_file.path).casecmp?('.zip') && !error
-      Zip::File.open(upload_file.path) do |zipfile|
-        Assignment.transaction do
-          # Build assignment from properties
-          prop_file = zipfile.find_entry('properties.yml')
-          assignment = build_uploaded_assignment(prop_file)
-          zipfile.remove(prop_file)
-          # Build peer review assignment if it exists
-          child_prop_file = zipfile.find_entry('peer-review-config-files/properties.yml')
-          unless child_prop_file.nil?
-            child_assignment = build_uploaded_assignment(child_prop_file, assignment)
-            child_assignment.save!
-            zipfile.remove(child_prop_file)
-          end
-          zipfile.each do |entry|
-            flash_message(:warning, I18n.t('assignments.unexpected_file_found', item: entry.name))
-          end
-          assignment.save!
-          redirect_to edit_assignment_path(assignment.id)
-        rescue StandardError => e
-          error = true
-          flash_message(:error, e.message)
-          raise ActiveRecord::Rollback
+    raise I18n.t('upload_errors.blank') if upload_file.size == 0
+    raise I18n.t('upload_errors.invalid_file_type', type: 'zip') unless File.extname(upload_file.path).casecmp?('.zip')
+
+    Zip::File.open(upload_file.path) do |zipfile|
+      ApplicationRecord.transaction do
+        # Build assignment from properties
+        prop_file = zipfile.get_entry(CONFIG_FILES[:properties])
+        assignment = build_uploaded_assignment(prop_file)
+        zipfile.remove(prop_file)
+        # Build peer review assignment if it exists
+        child_prop_file = zipfile.find_entry(CONFIG_FILES[:peer_review_properties])
+        unless child_prop_file.nil?
+          child_assignment = build_uploaded_assignment(child_prop_file, assignment)
+          child_assignment.save!
+          zipfile.remove(child_prop_file)
+        end
+        assignment.save!
+        zipfile.each do |entry|
+          flash_message(:warning, I18n.t('assignments.unexpected_file_found', item: entry.name))
         end
+        redirect_to edit_assignment_path(assignment.id)
       end
-    elsif !error
-      error = true
-      flash_message(:error, I18n.t('upload_errors.invalid_file_type', type: 'zip'))
     end
-    if error && params[:is_scanned] == 'true'
+  rescue StandardError => e
+    flash_message(:error, e.message)
+    if params[:is_scanned] == 'true'
       redirect_to new_assignment_path(scanned: true)
-    elsif error && params[:is_timed] == 'true'
+    elsif params[:is_timed] == 'true'
       redirect_to new_assignment_path(timed: true)
-    elsif error
+    else
       redirect_to new_assignment_path
     end
   end
 
   private
 
+  # Ensure that the +assignment+ type (scanned, timed, neither) matches the params
+  # If it does not match, raise an error
+  def check_assignment_type_match!(assignment)
```

Comment:
I suggest moving this to a separate function because the functionality is self contained and it keeps the build_uploaded_assignment method from getting cluttered